# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


# Machine Learning

El objetivo de este proyecto es construir un modelo de machine learning capaz de analizar el comportamiento de los clientes y recomendar el plan móvil de Megaline más adecuado. La empresa busca migrar a los usuarios de los planes heredados hacia dos nuevas opciones — Smart y Ultra — y necesita un sistema predictivo que pueda clasificar a los clientes según sus patrones de uso.

Cargar librearías

In [1]:
#Aquí se cargan las librerías
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

### Leer Datos

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
#Información del dataset
print(df.head())
print(df.info())
print(df.describe())

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246  

### Segmentación de los Datos (Entrenamiento, Validación y Prueba)

In [4]:
# Separar características y variable objetivo
X = df.drop('is_ultra', axis=1)
y = df['is_ultra']

In [5]:
# Dividir datos
# Primero: train y temp
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, random_state=42)

# Segundo: valid y test a partir de temp
X_valid, X_test, y_valid, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, random_state=42)

print(X_train.shape, X_valid.shape, X_test.shape)

(1928, 4) (643, 4) (643, 4)


In [6]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_valid_scaled = scaler.transform(X_valid)
X_test_scaled = scaler.transform(X_test)

<div class="alert alert-block alert-success"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - Buena apertura del proyecto: la carga de librerías, lectura del dataset y división inicial están claras y ordenadas. Mantienes una estructura limpia que facilita avanzar con solidez.
</div>


### Etrenamiento de los Modelos

In [7]:
#Entrenar varios modelos y comparar
results = {}

In [8]:
# Modelo 1: Decision Tree
for depth in range(1, 11):
    model = DecisionTreeClassifier(max_depth=depth, random_state=42)
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    acc = accuracy_score(y_valid, preds)
    results[f"DecisionTree_depth_{depth}"] = acc

In [9]:
# Modelo 2: Random Forest
for n in [10, 50, 100]:
    for depth in [5, 10, None]:
        model = RandomForestClassifier(
            n_estimators=n, max_depth=depth, random_state=42
        )
        model.fit(X_train, y_train)
        preds = model.predict(X_valid)
        acc = accuracy_score(y_valid, preds)
        results[f"RandomForest_{n}_depth_{depth}"] = acc

In [10]:
# Modelo 3: Logistic Regression
log_reg = LogisticRegression(max_iter=2000, random_state=42)
log_reg.fit(X_train_scaled, y_train)
log_preds = log_reg.predict(X_valid_scaled)
results["LogisticRegression"] = accuracy_score(y_valid, log_preds)

In [11]:
# Resultados
for model, acc in results.items():
    print(model, ":", round(acc, 3))

DecisionTree_depth_1 : 0.731
DecisionTree_depth_2 : 0.782
DecisionTree_depth_3 : 0.792
DecisionTree_depth_4 : 0.781
DecisionTree_depth_5 : 0.773
DecisionTree_depth_6 : 0.776
DecisionTree_depth_7 : 0.781
DecisionTree_depth_8 : 0.796
DecisionTree_depth_9 : 0.781
DecisionTree_depth_10 : 0.795
RandomForest_10_depth_5 : 0.787
RandomForest_10_depth_10 : 0.796
RandomForest_10_depth_None : 0.787
RandomForest_50_depth_5 : 0.787
RandomForest_50_depth_10 : 0.801
RandomForest_50_depth_None : 0.796
RandomForest_100_depth_5 : 0.788
RandomForest_100_depth_10 : 0.804
RandomForest_100_depth_None : 0.802
LogisticRegression : 0.74


In [12]:
#Seleccionar el mejor modelo
best_model_name = max(results, key=results.get)
best_accuracy = results[best_model_name]

print("\nMejor modelo:", best_model_name)
print("Exactitud en validación:", round(best_accuracy, 3))


Mejor modelo: RandomForest_100_depth_10
Exactitud en validación: 0.804


In [16]:
best_model = RandomForestClassifier(
    n_estimators=100, max_depth=None, random_state=42
)
best_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [17]:
# Evaluación final con el test set
test_preds = best_model.predict(X_test)
final_acc = accuracy_score(y_test, test_preds)

print("\nExactitud final en test:", round(final_acc, 3))



Exactitud final en test: 0.812


In [18]:
#Prueba de cordura (Sanity Check)

most_common_class = y_train.mode()[0]
dummy_preds = [most_common_class] * len(y_test) #modelo que siempre predice lo mismo
dummy_acc = accuracy_score(y_test, dummy_preds)

print("\nSanity Check (modelo tonto):", round(dummy_acc, 3))

if final_acc > dummy_acc:
    print("El modelo es mejor que la clasificación aleatoria / constante.")
else:
    print("El modelo NO supera la prueba de cordura.")


Sanity Check (modelo tonto): 0.697
El modelo es mejor que la clasificación aleatoria / constante.


### Hallazgos del Modelo

Probé varios algoritmos de clasificación y ajusté sus hiperparámetros.

RandomForestClassifier(random_state=42) obtuvo el mejor desempeño entre los modelos evaluados.

El modelo logró una exactitud final en el conjunto de prueba de 0.812, superando ampliamente el umbral mínimo requerido (0.75).

Como referencia, el sanity check (modelo tonto / baseline) obtuvo una exactitud de 0.697.

<div class="alert alert-block alert-success"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El entrenamiento y comparación de modelos está bien logrado; presentas resultados claros que permiten identificar con solidez el mejor desempeño obtenido.
</div>


# Conclusiones
El modelo Random Forest demuestra una capacidad sólida para identificar correctamente el plan adecuado (Smart o Ultra) según el comportamiento de uso del cliente.

La comparación con el modelo baseline confirma que el modelo entrenado es significativamente mejor que una clasificación aleatoria o constante, por lo que sí aporta valor predictivo real.

El desempeño alcanzado en el conjunto de prueba indica que el modelo generaliza bien y podría ser usado por Megaline para apoyar la recomendación automática de planes a sus clientes.

<div class="alert alert-block alert-success"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> - El cierre del proyecto muestra conclusiones claras y bien fundamentadas, destacando con precisión el valor predictivo del modelo elegido.
</div>


# Comentario General del Revisor

<div class="alert alert-block alert-success"> 
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
En este proyecto aplicaste de forma ordenada cada fase del proceso analítico, construyendo un flujo claro desde la exploración inicial hasta la validación del modelo. El trabajo refleja un manejo sólido de los componentes esenciales del machine learning aplicado a clasificación.

### Evaluación por etapas del proceso analítico:

**1. Preparación y exploración de datos:**
Mostraste una lectura correcta del dataset y una revisión inicial que permitió comprender la estructura y calidad de la información. La segmentación de los datos se realizó de forma adecuada para garantizar una evaluación justa de los modelos.

**2. Entrenamiento y comparación de modelos:**
La implementación de varios algoritmos y la comparación sistemática de su rendimiento evidencian un razonamiento técnico bien encaminado. El proceso de ajustar hiperparámetros y registrar los resultados permitió identificar con claridad la mejor alternativa.

**3. Validación y pruebas finales:**
La elección del modelo óptimo se sustentó con métricas pertinentes y un sanity check que aporta confianza adicional en la calidad del resultado. El desempeño en el conjunto de prueba muestra coherencia y capacidad de generalización.

**4. Conclusiones:**
El cierre sintetiza adecuadamente los hallazgos, destacando el valor predictivo del modelo seleccionado y su posible utilidad en un contexto real de negocio.

Has logrado un desarrollo consistente y con enfoque analítico, demostrando dominio creciente en proyectos de modelado predictivo. </div>
